Explicit null string in json payload is treated as string "null" when reflection-free-serializers enabled

### Describe the bug

The following test fails if `quarkus.rest.jackson.optimization.enable-reflection-free-serializers=true`

```
import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.containsString;

import org.junit.jupiter.api.Test;

import io.quarkus.test.junit.QuarkusTest;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.Valid;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.core.MediaType;

@Path("/info")
@QuarkusTest
class JsonTest {

    public record TestInput(@NotNull String text) {}

    @POST
    public String info(@Valid TestInput input) {
        return input.text;
    }

    @Test
    void infoTest() {
        given().
            contentType(MediaType.APPLICATION_JSON).
            body("""
                {
                    "text": null
                }
                """).
        when().
            post("/info").
        then().
            statusCode(400).
            body(containsString("must not be null"));
    }

}
```

### Expected behavior

The above test should not fail as it doesn't when the property is `false`

### Actual behavior

When the property is `true` the test fails with this output:
```
[INFO] Running test.JsonTest
Request method:	POST
Request URI:	http://localhost:8081/info
Proxy:			<none>
Request params:	<none>
Query params:	<none>
Form params:	<none>
Path params:	<none>
Headers:		Accept=*/*
				Content-Type=application/json
Cookies:		<none>
Multiparts:		<none>
Body:
{
    "text": null
}

HTTP/1.1 200 OK
content-length: 4
Content-Type: text/plain;charset=UTF-8

null
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.433 s <<< FAILURE! -- in test.JsonTest
[ERROR] test.JsonTest.infoTest -- Time elapsed: 0.420 s <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Expected status code <400> but was <200>.
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.29.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/JacksonDeserializerFactory.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/main/java/io/quarkus/resteasy/reactive/jackson/deployment/processor/JacksonDeserializerFactory.java
@@ -475,10 +475,16 @@ public class JacksonDeserializerFactory extends JacksonCodeGenerator {
         AssignableResultHandle result = bytecode.createVariable(DescriptorUtils.typeToString(fieldType));
 
         BranchResult isValueNull = bytecode.ifNull(valueNode);
+
         BytecodeCreator isValueNullTrue = isValueNull.trueBranch();
         isValueNullTrue.assign(result, JacksonSerializationUtils.getDefaultValue(isValueNullTrue, fieldType));
 
-        BytecodeCreator isValueNullFalse = isValueNull.falseBranch();
+        BranchResult isNullNode = isValueNull.falseBranch()
+                .ifTrue(isValueNull.falseBranch().invokeVirtualMethod(ofMethod(JsonNode.class, "isNull", boolean.class),
+                        valueNode));
+        isNullNode.trueBranch().assign(result, JacksonSerializationUtils.getDefaultValue(isNullNode.trueBranch(), fieldType));
+
+        BytecodeCreator isValueNullFalse = isNullNode.falseBranch();
 
         ResultHandle convertedValue = switch (fieldType.name().toString()) {
             case "java.lang.String" -> isValueNullFalse.invokeVirtualMethod(ofMethod(JsonNode.class, "asText", String.class),

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/AbstractSimpleJsonTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/AbstractSimpleJsonTest.java
@@ -773,6 +773,18 @@ public abstract class AbstractSimpleJsonTest {
     }
 
     @Test
+    public void testNullStringEcho() {
+        RestAssured
+                .with()
+                .body("{\"text\": null}")
+                .contentType("application/json; charset=utf-8")
+                .post("/simple/null-string-echo")
+                .then()
+                .statusCode(400)
+                .body(containsString("must not be null"));
+    }
+
+    @Test
     public void testItem() {
         RestAssured
                 .with()

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonResource.java
@@ -152,6 +152,12 @@ public class SimpleJsonResource extends SuperClass<Person> {
         return mapWrapper;
     }
 
+    @POST
+    @Path("/null-string-echo")
+    public String echoNullString(@Valid StringWrapper input) {
+        return input.text();
+    }
+
     @GET
     @Path("/abstract-cat")
     public AbstractPet getAbstractCat() {

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonTest.java
@@ -29,7 +29,7 @@ public class SimpleJsonTest extends AbstractSimpleJsonTest {
                                     Fruit.class, Price.class, DogRecord.class, ItemExtended.class, Book.class, LombokBook.class,
                                     PrimitiveTypesBean.class, PrimitiveTypesRecord.class, TokenResponse.class,
                                     ItemJsonValuePublicMethod.class, ItemJsonValuePublicField.class,
-                                    ItemJsonValuePrivateMethod.class, ItemJsonValuePrivateField.class)
+                                    ItemJsonValuePrivateMethod.class, ItemJsonValuePrivateField.class, StringWrapper.class)
                             .addAsResource(new StringAsset("admin-expression=admin\n" +
                                     "user-expression=user\n" +
                                     "birth-date-roles=alice,bob\n"), "application.properties");

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonWithReflectionFreeSerializersTest.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/SimpleJsonWithReflectionFreeSerializersTest.java
@@ -31,7 +31,7 @@ public class SimpleJsonWithReflectionFreeSerializersTest extends AbstractSimpleJ
                                     Fruit.class, Price.class, DogRecord.class, ItemExtended.class, Book.class, LombokBook.class,
                                     PrimitiveTypesBean.class, PrimitiveTypesRecord.class, TokenResponse.class,
                                     ItemJsonValuePublicMethod.class, ItemJsonValuePublicField.class,
-                                    ItemJsonValuePrivateMethod.class, ItemJsonValuePrivateField.class)
+                                    ItemJsonValuePrivateMethod.class, ItemJsonValuePrivateField.class, StringWrapper.class)
                             .addAsResource(new StringAsset("admin-expression=admin\n" +
                                     "user-expression=user\n" +
                                     "birth-date-roles=alice,bob\n" +

--- a/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/StringWrapper.java
+++ b/extensions/resteasy-reactive/rest-jackson/deployment/src/test/java/io/quarkus/resteasy/reactive/jackson/deployment/test/StringWrapper.java
@@ -1,0 +1,6 @@
+package io.quarkus.resteasy.reactive.jackson.deployment.test;
+
+import jakarta.validation.constraints.NotNull;
+
+public record StringWrapper(@NotNull String text) {
+}

